File: BiosenseSignal_Flutter_SDK_5.6.3/android/src/main/kotlin/com/biosensesignal/flutter_sdk/SessionManager.kt
package com.biosensesignal.flutter_sdk

import com.biosensesignal.sdk.api.HealthMonitorException
import com.biosensesignal.sdk.api.SessionEnabledVitalSigns
import com.biosensesignal.sdk.api.alerts.ErrorData
import com.biosensesignal.sdk.api.alerts.WarningData
import com.biosensesignal.sdk.api.images.DeviceOrientation
import com.biosensesignal.sdk.api.images.ImageData
import com.biosensesignal.sdk.api.images.ImageListener
import com.biosensesignal.sdk.api.license.LicenseDetails
import com.biosensesignal.sdk.api.license.LicenseInfo
import com.biosensesignal.sdk.api.ppg_device_scanner.PPGDevice
import com.biosensesignal.sdk.api.ppg_device_scanner.PPGDeviceInfo
import com.biosensesignal.sdk.api.ppg_device_scanner.PPGDeviceScanner
import com.biosensesignal.sdk.api.ppg_device_scanner.PPGDeviceScannerListener
import com.biosensesignal.sdk.api.session.Session
import com.biosensesignal.sdk.api.session.SessionInfoListener
import com.biosensesignal.sdk.api.session.SessionState
import com.biosensesignal.sdk.api.session.demographics.Sex
import com.biosensesignal.sdk.api.session.demographics.SubjectDemographic
import com.biosensesignal.sdk.api.session.ppg_device.PPGDeviceInfoListener
import com.biosensesignal.sdk.api.vital_signs.VitalSign
import com.biosensesignal.sdk.api.vital_signs.VitalSignsListener
import com.biosensesignal.sdk.api.vital_signs.VitalSignsResults
import com.biosensesignal.sdk.ppg_data.ppg_device.PPGDeviceType
import com.biosensesignal.sdk.ppg_device_scanner.PPGDeviceScannerFactory
import com.biosensesignal.sdk.session.FaceSessionBuilder
import com.biosensesignal.sdk.session.PolarSessionBuilder
import android.content.Context
import com.biosensesignal.sdk.api.fall_detection.FallDetectionData
import com.biosensesignal.sdk.api.fall_detection.FallDetectionListener
import com.biosensesignal.sdk.api.session.CameraLocation
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow


class SessionManager(private val eventChannel: BiosenseSignalEventChannel):
    ImageDataSource, ImageListener, VitalSignsListener, SessionInfoListener,
    PPGDeviceInfoListener, FallDetectionListener {

    private val _images = MutableSharedFlow<ImageData>(replay = 0, extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_LATEST)
    private var session: Session? = null

    private val ppgScanners = mutableMapOf<String, PPGDeviceScanner>()
    override val images: Flow<ImageData> = _images

    @Throws(HealthMonitorException::class)
    fun createCameraSession(
        context: Context,
        licenseKey: String,
        productId: String? = null,
        deviceOrientation: Int? = null,
        subjectSex: Int? = null,
        subjectAge: Double? = null,
        subjectWeight: Double? = null,
        subjectHeight: Double? = null,
        detectionAlwaysOn: Boolean? = false,
        strictMeasurementGuidance: Boolean? = false,
        sdkAnalytics: Boolean? = false,
        cameraLocation: Int? = null,
        options: Map<String, Any>? = null
    ) {
        val orientation = resolveDeviceOrientation(deviceOrientation)
        val subjectDemographic = resolveSubjectDemographic(
            subjectSex,
            subjectAge,
            subjectWeight,
            subjectHeight
        )

        val sessionBuilder = FaceSessionBuilder(context)
            .withDeviceOrientation(orientation)
            .withSubjectDemographic(subjectDemographic)

        detectionAlwaysOn?.let {
            sessionBuilder.withDetectionAlwaysOn(it)
        }

        strictMeasurementGuidance?.let {
            sessionBuilder.withStrictMeasurementGuidance(it)
        }

        resolveCameraLocation(cameraLocation)?.let { location ->
            sessionBuilder.withCameraLocation(location)
        }

        session = sessionBuilder
            .withImageListener(this@SessionManager)
            .withVitalSignsListener(this@SessionManager)
            .withSessionInfoListener(this@SessionManager)
            .let { if (sdkAnalytics == true) it.withAnalytics() else it }
            .withOptions(options)
            .build(LicenseDetails(licenseKey, productId))
    }

    @Throws(HealthMonitorException::class)
    fun createPPGDeviceSession(
        context: Context,
        licenseKey: String,
        productId: String? = null,
        deviceId: String,
        deviceType: Int,
        subjectSex: Int? = null,
        subjectAge: Double? = null,
        subjectWeight: Double? = null,
        subjectHeight: Double? = null,
        fallDetection: Boolean? = false,
        sdkAnalytics: Boolean? = false,
        options: Map<String, Any>? = null
    ) {
        val subjectDemographic = resolveSubjectDemographic(
            subjectSex,
            subjectAge,
            subjectWeight,
            subjectHeight
        )

        if (resolvePPGDeviceType(deviceType) == PPGDeviceType.POLAR) {
            var builder = PolarSessionBuilder(context, deviceId)
                .withSubjectDemographic(subjectDemographic)
                .withVitalSignsListener(this@SessionManager)
                .withSessionInfoListener(this@SessionManager)
                .withPPGDeviceInfoListener(this)
                .let { if (sdkAnalytics == true) it.withAnalytics() else it }
                .withOptions(options)

            if (fallDetection == true) {
                builder = builder.withFallDetectionListener(this)
            }

            session = builder.build(LicenseDetails(licenseKey, productId))
        }
    }

    fun startPPGDevicesScan(
        context: Context,
        scannerId: String,
        deviceType: Int,
        timeout: Long?
    ) {
        when (resolvePPGDeviceType(deviceType)) {
            PPGDeviceType.POLAR -> {
                ppgScanners[scannerId] = PPGDeviceScannerFactory.create(
                    context,
                    PPGDeviceType.POLAR,
                    object : PPGDeviceScannerListener {
                        override fun onPPGDeviceDiscovered(device: PPGDevice) {
                            eventChannel.sendEvent(
                                NativeBridgeEvents.ppgDeviceDiscovered,
                                mapOf(
                                    Pair("scannerId", scannerId),
                                    Pair("device", device.toMap()),
                                )
                            )
                        }

                        override fun onPPGDeviceScanFinished() {
                            eventChannel.sendEvent(NativeBridgeEvents.ppgDeviceScanFinished, scannerId)
                        }
                    }
                ).also { scanner ->
                    timeout?.let {
                        scanner.start(timeout)
                    } ?: scanner.start()
                }
            }
            else -> return
        }
    }

    fun stopPPGDeviceScan(scannerId: String) {
        ppgScanners[scannerId]?.stop();
    }

    @Throws(HealthMonitorException::class)
    fun startSession(duration: Int?) {
        session?.start(duration?.toLong() ?: 0)
    }

    @Throws(HealthMonitorException::class)
    fun stopSession() {
        session?.stop()
    }

    fun terminateSession() {
        session?.terminate()
    }

    fun getSessionState(): SessionState? {
        return session?.state
    }

    override fun onVitalSign(vitalSign: VitalSign) {
        eventChannel.sendEvent(NativeBridgeEvents.sessionVitalSign, vitalSign.toMap() ?: return)
    }

    override fun onFinalResults(vitalSignsResults: VitalSignsResults) {
        val results = vitalSignsResults.results.mapNotNull { result ->
            result.toMap()
        }
        eventChannel.sendEvent(NativeBridgeEvents.sessionFinalResults, results)
    }

    override fun onImage(imageData: ImageData) {
        eventChannel.sendEvent(NativeBridgeEvents.imageData, imageData.toMap())
        _images.tryEmit(imageData)
    }

    override fun onSessionStateChange(sessionState: SessionState) {
        eventChannel.sendEvent(NativeBridgeEvents.sessionStateChange, sessionState.ordinal)
    }

    override fun onWarning(warningData: WarningData) {
        eventChannel.sendEvent(NativeBridgeEvents.sessionWarning, warningData.toMap())
    }

    override fun onError(errorData: ErrorData) {
        eventChannel.sendEvent(NativeBridgeEvents.sessionError, errorData.toMap())
    }

    override fun onLicenseInfo(licenseInfo: LicenseInfo) {
        eventChannel.sendEvent(NativeBridgeEvents.licenseInfo, licenseInfo.toMap())
    }

    override fun onEnabledVitalSigns(enabledVitalSigns: SessionEnabledVitalSigns) {
        eventChannel.sendEvent(NativeBridgeEvents.enabledVitalSigns, enabledVitalSigns.toMap())
    }

    override fun onPPGDeviceInfo(ppgDeviceInfo: PPGDeviceInfo) {
        eventChannel.sendEvent(NativeBridgeEvents.ppgDeviceInfo, ppgDeviceInfo.toMap())
    }

    override fun onPPGDeviceBatteryLevel(level: Int) {
        eventChannel.sendEvent(NativeBridgeEvents.ppgDeviceBattery, level)
    }

    override fun onFallDetectionData(data: FallDetectionData) {
        eventChannel.sendEvent(NativeBridgeEvents.fallDetectionData, data.toMap())
    }

    private fun resolveDeviceOrientation(deviceOrientation: Int?): DeviceOrientation? {
        return deviceOrientation?.let {orientation ->
            try {
                DeviceOrientation.values()[orientation]
            } catch (ignore: IndexOutOfBoundsException) {
                null
            }
        }
    }

    private fun resolveSubjectDemographic(sexInt: Int?, age: Double?, weight: Double?, height: Double?): SubjectDemographic {
        val sex = sexInt?.let { sex ->
            try {
                Sex.values()[sex]
            } catch (e: IndexOutOfBoundsException) {
                Sex.UNSPECIFIED
            }
        }

        return SubjectDemographic(sex, age, weight, height)
    }

    private fun resolvePPGDeviceType(ppgDeviceType: Int?): PPGDeviceType {
        if (ppgDeviceType == PPGDeviceType.POLAR.ordinal) {
            return PPGDeviceType.POLAR
        }

        return PPGDeviceType.POLAR
    }

    private fun resolveCameraLocation(cameraLocation: Int?): CameraLocation? {
        return cameraLocation?.let { location ->
            try {
                CameraLocation.values()[location]
            } catch (ignore: IndexOutOfBoundsException) {
                null
            }
        }
    }
}

File: BiosenseSignal_Flutter_SDK_5.6.3/ios/Classes/SessionManager.swift

import Foundation
import Combine

class SessionManager:
        ImageDataSource,
        ImageListener,
        VitalSignsListener,
        SessionInfoListener,
        PPGDeviceInfoListener,
        FallDetectionListener {
    
    static let shared = SessionManager()
    
    var eventChannel: BiosenseSignalEventChannel?
    private var session: Session?
    private var ppgScanners = [String: PPGDeviceScanner]()
    
    let images = PassthroughSubject<ImageData, Never>()
    
    private init() {}
    
    func createSession(
        licenseKey: String,
        productId: String? = nil,
        deviceOrientation: Int? = nil,
        subjectSex: Int? = nil,
        subjectAge: Double? = nil,
        subjectWeight: Double? = nil,
        subjectHeight: Double? = nil,
        detectionAlwaysOn: Bool? = false,
        imageFormatMode: Int? = nil,
        strictMeasurementGuidance: Bool? = false,
        sdkAnalytics: Bool? = false,
        cameraLocation: Int? = nil,
        options: [String: Any]? = nil
    
    ) throws {
        var sessionBuilder = FaceSessionBuilder()
        if let subjectDemographic = resolveSubjectDemographic(sex: subjectSex, age: subjectAge, weight: subjectWeight, height: subjectHeight) {
            sessionBuilder = sessionBuilder.withSubjectDemographic(subjectDemographic)
        }
        
        if let orientation = resolveDeviceOrientation(deviceOrientation: deviceOrientation) {
            sessionBuilder = sessionBuilder.withDeviceOrientation(orientation)
        }

        if let cameraLocation = resolveCameraLocation(cameraLocation: cameraLocation) {
            sessionBuilder = sessionBuilder.withCameraLocation(cameraLocation)
        }
        
        if let imageFormat = resolveImageFormat(imageFormatMode: imageFormatMode) {
            sessionBuilder = sessionBuilder.withImageFormatMode(imageFormat)
        }

        if let strictMeasurementGuidance = strictMeasurementGuidance {
            sessionBuilder = sessionBuilder.withStrictMeasurementGuidance(strictMeasurementGuidance)
        }

        if let detectionAlwaysOn = detectionAlwaysOn {
            sessionBuilder = sessionBuilder.withDetectionAlwaysOn(detectionAlwaysOn)
        }
        
        if (sdkAnalytics == true) {
            sessionBuilder = sessionBuilder.withAnalytics() as! FaceSessionBuilder
        }
        
        session = try sessionBuilder
            .withImageListener(self)
            .withVitalSignsListener(self)
            .withSessionInfoListener(self)
            .withOptions(options: options ?? [:])
            .build(licenseDetails: LicenseDetails(licenseKey: licenseKey, productId: productId))
    }
    
    func createPPGDeviceSession(
        licenseKey: String,
        productId: String? = nil,
        deviceId: String,
        deviceType: Int,
        subjectSex: Int? = nil,
        subjectAge: Double? = nil,
        subjectWeight: Double? = nil,
        subjectHeight: Double? = nil,
        fallDetection: Bool? = false,
        sdkAnalytics: Bool? = false,        
        options: [String: Any]? = nil) throws {
            if (resolveDeviceType(deviceType: deviceType) != PPGDeviceType.polar) {
                throw NSError(domain: AlertDomains.initialization, code: AlertCodes.ppgDeviceUnsupportedDeviceModelError)
            }
            
            var sessionBuilder = PolarSessionBuilder(polarDeviceID: deviceId)
            if let subjectDemographic = resolveSubjectDemographic(sex: subjectSex, age: subjectAge, weight: subjectWeight, height: subjectHeight) {
                sessionBuilder = sessionBuilder.withSubjectDemographic(subjectDemographic)
            }
            
            if (fallDetection == true) {
                sessionBuilder = sessionBuilder.withFallDetectionListener(self)
            }
            
            if (sdkAnalytics == true) {
                sessionBuilder = sessionBuilder.withAnalytics() as! PolarSessionBuilder
            }
            
            session = try sessionBuilder
                .withVitalSignsListener(self)
                .withSessionInfoListener(self)
                .withPPGDeviceInfoListener(self)
                .withOptions(options: options ?? [:])
                .build(licenseDetails: LicenseDetails(licenseKey: licenseKey, productId: productId))
        }
    
    func startPPGDevicesScan(scannerId: String, deviceType: Int, timeout: Int?) throws {
        if (resolveDeviceType(deviceType: deviceType) != PPGDeviceType.polar) {
            return
        }

        let ppgScanListener = PPGDeviceScannerListenerImpl(eventChannel: eventChannel, scannerId: scannerId)
        let scanner = try PPGDeviceScannerFactory.create(ppgDeviceType: PPGDeviceType.polar, listener: ppgScanListener)
        
        ppgScanners[scannerId] = scanner
        if let timeout = timeout {
            try scanner.start(timeout: UInt(timeout))
        } else {
            try scanner.start()
        }
    }
    
    func stopPPGDevicesScan(scannerId: String) {
        ppgScanners[scannerId]?.stop()
    }
    
    func startSession(duration: Int?) throws {
        try session?.start(measurementDuration: UInt64(duration ?? 0))
    }

    func stopSession() throws {
        try session?.stop()
    }

    func terminateSession() {
        session?.terminate()
    }

    func getSessionState() -> SessionState? {
        return session?.state
    }

    func onImage(imageData: ImageData) {
        images.send(imageData)
        eventChannel?.sendEvent(name: NativeBridgeEvents.imageData , payload: imageData.toMap())
    }
    
    func onVitalSign(vitalSign: VitalSign) {
        if let map = vitalSign.toMap() {
            eventChannel?.sendEvent(name: NativeBridgeEvents.sessionVitalSign , payload: map)
        }
    }
    
    func onFinalResults(results: VitalSignsResults) {
        let finalResults = results.getResults().compactMap { result in
            result.toMap()
        }
        
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionFinalResults, payload: finalResults)
    }
    
    func onSessionStateChange(sessionState: SessionState) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionStateChange, payload: sessionState.rawValue)
    }
    
    func onWarning(warningData: WarningData) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionWarning, payload: warningData.toMap())
    }
    
    func onError(errorData: ErrorData) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionError, payload: errorData.toMap())
    }
    
    func onLicenseInfo(licenseInfo: LicenseInfo) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.licenseInfo, payload: licenseInfo.toMap())
    }
    
    func onEnabledVitalSigns(enabledVitalSigns: SessionEnabledVitalSigns) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.enabledVitalSigns, payload: enabledVitalSigns.toMap())
    }
    
    func onPPGDeviceBatteryLevel(_ batteryLevel: UInt) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.ppgDeviceBatteryLevel, payload: batteryLevel)
    }
    
    func onPPGDeviceInfo(_ info: BiosenseSignal.PPGDeviceInfo) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.ppgDeviceInfo, payload: info.toMap())
    }
    
    func onFallDetectionData(_ data: FallDetectionData) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.fallDetectionData, payload: data.toMap())
    }

    private func resolveDeviceOrientation(deviceOrientation: Int?) -> DeviceOrientation? {
        guard let orientation = deviceOrientation, let orientationEnum = DeviceOrientation.init(rawValue: orientation) else {
            return nil
        }
        
        return orientationEnum
    }

    private func resolveSubjectDemographic(sex: Int?, age: Double?, weight: Double?, height: Double?) -> SubjectDemographic? {
        if (sex == nil && age == nil && weight == nil && height == nil) {
            return nil
        }
        
        var sdkAge: NSNumber?
        if let age = age {
            sdkAge = NSNumber.init(value: age)
        }
        
        var sdkWeight: NSNumber?
        if let weight = weight {
            sdkWeight = NSNumber.init(value: weight)
        }
        
        var sdkHeight: NSNumber?
        if let height = height {
            sdkHeight = NSNumber.init(value: height)
        }
        
        return SubjectDemographic(
            sex: Sex.init(rawValue: sex ?? 0) ?? Sex.unspecified,
            age: sdkAge,
            weight: sdkWeight,
            height: sdkHeight
        )
    }

    private func resolveImageFormat(imageFormatMode: Int?) -> ImageFormatMode? {
        guard let imageFormat = imageFormatMode,
              let imageFormatEnum = ImageFormatMode.init(rawValue: imageFormat) else {
            return nil
        }

        return imageFormatEnum
    }
    
    private func resolveDeviceType(deviceType: Int) -> PPGDeviceType? {
        return PPGDeviceType.init(rawValue: deviceType)
    }
    
    private func resolveCameraLocation(cameraLocation: Int?) -> CameraLocation? {
        guard let cameraLocation = cameraLocation,
              let cameraLocationEnum = CameraLocation.init(rawValue: cameraLocation) else {
            return nil
        }

        return cameraLocationEnum
    }
}

extension SessionManager {
    
    class PPGDeviceScannerListenerImpl: PPGDeviceScannerListener {
        var eventChannel: BiosenseSignalEventChannel?
        let scannerId: String
        
        
        init(eventChannel: BiosenseSignalEventChannel?, scannerId: String) {
            self.eventChannel = eventChannel
            self.scannerId = scannerId
        }
        
        func onPPGDeviceDiscovered(ppgDevice: PPGDevice) {
            let device = ppgDevice.toMap()
            let result: [String: Any] = [
                "scannerId": scannerId,
                "device": device
            ]
            eventChannel?.sendEvent(name: NativeBridgeEvents.ppgDeviceDiscovered, payload: result)
        }
        
        func onPPGDeviceScanFinished() {
            eventChannel?.sendEvent(name: NativeBridgeEvents.ppgDeviceScanFinished, payload: scannerId)
        }
    }
}
